### DISCUSSION TITLE
Lazy initialization of observers

### DISCUSSION THREAD

**Discussion by laidinidis** [robfallows/tunguska-reactive-aggregate#53] [Q&A]
Is there a way to initialize observers lazily? 
When I'm doing 
``` javascript
ReactiveAggregate(this, Collection, [
    { $match: selectors },
    { $group: { ... } }
  ], {
  debounceCount: 100,
  debounceDelay: 1000,
  noAutomaticObserver: true,
}
```
the pipeline runs really fast, but if I add an observer it's slow
``` javascript
ReactiveAggregate(this, Collection, [
    { $match: selectors },
    { $group: { ... }}
  ], {
  debounceCount: 100,
  debounceDelay: 1000,
  noAutomaticObserver: true,
  observers: [
    Collection.find(selectors)
  ]
}
```
What I want is to run the pipeline, get the results and then initialize the observers.

**Comment by robfallows** [comment]
@laidinidis : I converted this issue to a Q&A discussion.

Great question!

It should be possible to wrap the observer setup in a `Meter.defer` or `Meteor.setTimeout` to make that asynchronous. Everything else should catch up after that. That will need testing because it's untried.

Note that in your first example, the pipeline will not re-run reactively since you do not specify any `observers` and you have `noAutomaticObserver: true`. That is basically a "run once" configuration. Using a `debounceDelay: 1000` value will alter the behaviour to "run every second".

At the moment my time is limited, so if you think you can help with a PR, that would be great. Otherwise, I'll find some time over the next couple of weeks to test.

**Reply by laidinidis** [reply]
I didn't try `Meteor.defer` or `Meteor.setTimeout`, I'll give it a try.
I know that by not specifying `observers` the pipeline will run only once however using debounce didn't re-run it every second.
I might find some time to open a pr.
Thanks for the response

**Reply by robfallows** [reply]
> I didn't try Meteor.defer or Meteor.setTimeout, I'll give it a try.

Just to clarify - that change is in the package code.

I've tested that the change doesn't break the package, but whether it will improve your initialising of observer performance, I don't know.

If you want to test, you will need to make a local clone of the repo in a `packages/` folder under your project's root directory. Clone the repo and switch to the `asynchronous-observer-setup` branch.

```text
# From your project root:
mkdir packages  # If doesn't already exist
cd packages
git clone git@github.com:robfallows/tunguska-reactive-aggregate.git
cd tunguska-reactive-aggregate
git checkout asynchronous-observer-setup
cd ../..
```

You should then rebuild your project. If all went well you should see v1.3.5 of `tunguska-reactive-aggregate` being installed. Please test and report back!

**Reply by laidinidis** [reply]
> Just to clarify - that change is in the package code.

> If you want to test, you will need to make a local clone of the repo in a packages/ folder under your project's root directory.

Yes, I know how to do that :) 
I'll try the `asynchronous-observer-setup` branch tomorrow.
Thanks for the help, really appreciate it!

**Reply by laidinidis** [reply]
I can confirm that the change does not break anything and it also improves observer init time!

**Reply by robfallows** [reply]
Awesome! I'll try and get round to issuing a new release later today.

Thanks for the confirmation 🙂

**Reply by robfallows** [reply]
New release now published.

**Reply by robfallows** [reply]
@laidinidis : Please see #55, which is showing poor performance from asynchronous observer initialisation.

I wonder if you can run a similar test with `debug: true` in the options and paste the console output here, please?

**Reply by robfallows** [reply]
@laidinidis : There have been some reports of serious performance issues with v1.3.5 and I have reverted the changes for asynchronous observer set up for v1.3.6. If v1.3.5 is working for you without issue, then you can continue to use it for the forseeable future, but you should pin the version or any update will change behaviour.

**Reply by laidinidis** [reply]
Hi Rob, sorry I didn't reply earlier, I was on vacation. 
To be honest I didn't notice any performance issues or improvements on production.  I'm guessing i'll be fine with the new version. 
Thanks for the heads-up.
